### PR TITLE
RSPEC-1048 Destructors should not throw exceptions (For VB)

### PR DESCRIPTION
**Pull request by loic-joly-sonarsource**
This is not a direct translation, since VB does not have the notion of destructor, so review should be more careful.

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+using SonarAnalyzer.Rules.Common;
+
+namespace SonarAnalyzer.Rules.VisualBasic
+{
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotThrowFromDestructors : DoNotThrowFromDestructorsBase
+    {
+        private const string MessageFormat = "Remove this 'Throw' statement.";
+        private static readonly DiagnosticDescriptor rule =
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs]: [minor][cosmetic] Please add a blank line between the `const` and `static readonly`.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs]: [minor][cosmetic] un-needed blank line.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs]: [minor][cosmetic] we try (at least for new rules) to move the `c =>` to the next line.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                SyntaxKind.ThrowStatement);
+        }
+
+        private bool IsFinalizer(MethodBlockSyntax methodBlockSyntax)
+        {
+            if (methodBlockSyntax == null)
+            {
+                return false;
+            }
+            var decl = methodBlockSyntax.SubOrFunctionStatement;
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs]: [minor][cosmetic] please add a blank line after the `if` block.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs (hunk 3 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs]: Also please fix the indent of the lambda block.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            var noParam = decl.ParameterList == null || decl.ParameterList.Parameters.Count == 0;
+            var noTypeParam = decl.TypeParameterList == null || decl.TypeParameterList.Parameters.Count == 0;
+            var isSub = decl.SubOrFunctionKeyword.Kind() == SyntaxKind.SubKeyword;
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs]: Please use `IsKind()` instead of `Kind() ==`.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs (hunk 4 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs]: Also please use a full name and not `decl`.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            var isProtected = decl.Modifiers.Any(testc => testc.Kind() == SyntaxKind.ProtectedKeyword);
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs]: Same here, `IsKind()` instead of `Kind() ==`. Also I think there is an overload to `Any` or something like `AnyKind` which does the same.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/RspecStrings.resx [hunk 7]
```diff
@@ -1174,7 +1204,7 @@
     <value>Because it is easy to extract strings from a compiled application, credentials should never be hard-coded. Do so, and they're almost guaranteed to end up in the hands of an attacker. This is particularly true for applications that are distributed.</value>
   </data>
   <data name="S2068_IsActivatedByDefault" xml:space="preserve">
-    <value>True</value>
+    <value>False</value>
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/RspecStrings.resx]: This change should not be part of this PR.

**Comment by loic-joly-sonarsource** [comment]
@Evangelink (not sure why, I cannot directly answer your comment...)
> Are we handling the fact that an exception thrown is caught in the destructor?
No, we are not. This is consistent with the C# behaviour, with a specific unit test about that, that claims: 
```c#
 // Noncompliant, generally a bad idea to throw and catch in the same method
```

I did not feel a strong need to worry about that because finalizers should not be common, finalizers that throw even less (for instance, the rule is never triggered in integration tests), and finalizers that throw but then do it in a try/catch are a corner or a corner case...

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs (hunk 6 above)
Comment (loic-joly-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs]: I could not find `AnyKind` or anything similar...

Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs]: If you look at `Metrics.cs` line `129` you can see that there is a built-in extension method (for C#) called `Any` that takes directly a `SyntaxKind`. I don't know if such extension methods exists for VB.NET. If it doesn't then please ignore this comment.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs (hunk 4 above)
Comment (loic-joly-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs]: Is there a reference explaining the team logic about blank lines? Because you often ask to remove "unneeded" blank lines or to add blanks lines (that I would consider unneeded), and I fail to see where.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/RspecStrings.resx (hunk 7 above)
Comment (loic-joly-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/RspecStrings.resx]: Do you know how this can happen?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs (hunk 4 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotThrowFromDestructors.cs]: No we don't have one because we naturally had the same logic within the team (for spacing) so we didn't do anything.

**Comment by Evangelink** [comment]
@loic-joly-sonarsource regarding the resx change, there is a bug in the `rspec.ps1` script.

And regarding the throw/catch, I wasn't asking for a change of behavior. I just wanted to confirm that the unit test you wrote was what we were expecting :)
